packages/cli/src/lib/features.ts

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2020 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  isValidPackageFeatureType,
+  PackageRole,
+  type BackstagePackageFeature,
+  type BackstagePackageFeatureType,
+} from '@backstage/cli-node';
+import { Project, SyntaxKind, ts, Type } from 'ts-morph';
+import { resolve as resolvePath } from 'node:path';
+import { EntryPoint } from './entryPoints';
+
+// A list of the package roles we want to extract features for
+const targetPackageRoles: PackageRole[] = [
+  'backend-plugin',
+  'backend-plugin-module',
+  'frontend-plugin',
+  'frontend-plugin-module',
+];
+
+// A list of the feature types we want to extract from the project
+// and include in the metadata
+const targetFeatureTypes: BackstagePackageFeatureType[] = [
+  '@backstage/BackendFeature',
+  '@backstage/BackstagePlugin',
+  '@backstage/FrontendPlugin',
+  '@backstage/FrontendModule',
+];
+
+// Returns all valid Backstage package features from a project
+// for all entry points.
+export function getFeaturesMetadata(
+  project: Project,
+  role: PackageRole,
+  dir: string,
+  entryPoints: EntryPoint[],
+): BackstagePackageFeature[] {
+  if (!targetPackageRoles.includes(role)) {
+    return [];
+  }
+
+  return entryPoints
+    .flatMap(entryPoint => getEntryPointExports(entryPoint, project, dir))
+    .filter(isDefaultExport)
+    .filter(isTargetFeatureType)
+    .map(toBackstagePackageFeature);
+}
+
+type EntryPointExport = {
+  name: string;
+  location: string;
+  type: BackstagePackageFeatureType;
+};
+
+// Returns all exports (default and named) from an entry point
+// that are valid Backstage package features
+export function getEntryPointExports(
+  { mount: location, path, ext }: EntryPoint,
+  project: Project,
+  dir: string,
+): EntryPointExport[] {
+  const fullFilePath = resolvePath(dir, path);
+  const sourceFile = project.getSourceFile(fullFilePath);
```

Comment:
Depending on when we run this I realized that it might be better to run towards `dist-types` instead, that should be a lot faster than scanning the source files. I'm thinking we'd be able to make this part of the `package prepack` command at that point, so that you don't need the changes to `package.json` in the repo at all.

If it's simple I think it would be useful to do a quick speed comparison of running this towards `src` vs `dist-types`

Reply:
Thinking about this some more I realize that we have an even more simplified version of the types inside the `dist` dir itself. Which should always exist and be up to date during `prepack`. i.e. we try pointing this at e.g. `dist/index.d.ts` instead.